main.go

```diff
@@ -0,0 +1,367 @@
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		var nsk *types.NoSuchKey
+		if errors.As(err, &nsk) {
+			return nil, noSuchKey{}
+		}
+		return nil, fmt.Errorf("getting from bucket %q with key %q: %w", bucket, key, err)
+	}
+
+	var entries entries
+	gzipReader, err := gzip.NewReader(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("making gzipReader: %w", err)
+	}
+	err = cbor.NewDecoder(gzipReader).Decode(&entries)
+	if err != nil {
+		return nil, fmt.Errorf("reading body from bucket %q with key %q: %w", bucket, key, err)
+	}
+
+	if len(entries.Entries) != int(t.size) || t.end != t.start+t.size {
+		return nil, fmt.Errorf("internal inconsistency: len(entries) == %d; tile = %v", len(entries.Entries), t)
+	}
+
+	return &entries, nil
+}
+
+// tileCachingHandler is the main HTTP handler that serves CT tiles it fetches
+// from a backend server and from the cache tiles it maintains in S3.
+type tileCachingHandler struct {
+	logURL   string // The string form of the HTTP host and path prefix to add incoming request paths to in order to fetch tiles from the backing CT log. Must not be empty.
+	tileSize int    // The CT tile size used here and in the given backend. Must not be zero.
+
+	s3Service *s3.Client // The S3 service to use for caching tiles. Must not be nil.
+	s3Prefix  string     // The prefix to add to the path when caching tiles in S3. Must not be empty.
+	s3Bucket  string     // The S3 bucket to use for caching tiles. Must not be empty.
+}
+
+func (tch *tileCachingHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if !strings.HasSuffix(r.URL.Path, "/ct/v1/get-entries") {
+		w.WriteHeader(http.StatusNotFound)
+		fmt.Fprintf(w, "invalid path %q\n", r.URL.Path)
+		return
+	}
+
+	start, end, err := parseQueryParams(r.URL.Query())
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(w, err)
+		return
+	}
+
+	tile := makeTile(start, int64(tch.tileSize), tch.logURL, tch.s3Prefix)
+
+	contents, err := getFromS3(r.Context(), tch.s3Service, tch.s3Bucket, tile)
+	if err != nil && errors.Is(err, noSuchKey{}) {
+		contents, err = getTileFromBackend(r.Context(), r.URL.Path, tile)
+		if err != nil {
+			status := http.StatusInternalServerError
+			var statusCodeErr statusCodeError
+			if errors.As(err, &statusCodeErr) {
+				status = statusCodeErr.statusCode
+			}
+			w.WriteHeader(status)
+			fmt.Fprintln(w, err)
+			return
+		}
+
+		// If we got a partial tile, assume we are at the end of the log and the last
+		// tile isn't filled up yet. In that case, don't write to S3, but still return
+		// results to the user.
+		if len(contents.Entries) == tch.tileSize {
+			err := writeToS3(r.Context(), tch.s3Service, tch.s3Bucket, tile, contents)
+			if err != nil {
+				w.WriteHeader(http.StatusInternalServerError)
+				fmt.Fprintf(w, "writing to s3: %s\n", err)
+				return
+			}
+		} else {
+			w.Header().Set("X-Partial-Tile", "true")
+		}
+
+		w.Header().Set("X-Source", "CT log")
+	} else if err != nil {
```

Comment:
nit: I'd move this case above the getTileFromBackend case, so that the simple "just return the error" case isn't squashed between the "kinda-success" and "complete-success" cases.